client/src/unifycr.c

```diff
@@ -2914,29 +2732,151 @@ int unifycrfs_mount(const char prefix[], size_t size, int rank)
-    return unifycr_chunk_mem;
+    if (pcontext != NULL && *pcontext != NULL) {
+        /* define a temporary to refer to context */
+        unifycr_client_rpc_context_t* ctx = *pcontext;
+
+        /* free margo address to server */
+        margo_addr_free(ctx->mid, ctx->svr_addr);
+
+        /* shut down margo */
+        margo_finalize(ctx->mid);
+
+        /* free memory allocated for context structure,
+         * and set caller's pointer to NULL */
+        free(ctx);
+        *pcontext = NULL;
+    }
+
+    return UNIFYCR_SUCCESS;
 }
 
-/* get a list of chunks for a given file (useful for RDMA, etc.) */
-chunk_list_t* unifycr_get_chunk_list(char* path)
+/* free resources allocated during unifycr_init,
+ * generally we do this in reverse order that
+ * things were initailized in */
+static int unifycr_finalize(void)
 {
-    return NULL;
+    int rc = UNIFYCR_SUCCESS;
+
+    if (!unifycr_initialized) {
+        /* not initialized yet, so we shouldn't call finalize */
+        return UNIFYCR_FAILURE;
+    }
+
+    /* close spillover files */
+    if (unifycr_spilloverblock != 0) {
+        close(unifycr_spilloverblock);
+        unifycr_spilloverblock = 0;
+    }
+
+    if (unifycr_spillmetablock != 0) {
+        close(unifycr_spillmetablock);
+        unifycr_spillmetablock = 0;
+    }
+
+    /* detach from superblock */
+    unifycr_shm_free(shm_super_name, shm_super_size, &shm_super_buf);
+
+    /* free directory stream stack */
+    if (unifycr_dirstream_stack != NULL) {
+        free(unifycr_dirstream_stack);
+        unifycr_dirstream_stack = NULL;
+    }
+
+    /* free file stream stack */
+    if (unifycr_stream_stack != NULL) {
+        free(unifycr_stream_stack);
+        unifycr_stream_stack = NULL;
+    }
+
+    /* free file descriptor stack */
+    if (unifycr_fd_stack != NULL) {
+        free(unifycr_fd_stack);
+        unifycr_fd_stack = NULL;
+    }
+
+    /* no longer initialized, so update the flag */
+    unifycr_initialized = 0;
+
+    return rc;
+}
+
+/**
+ * unmount the mounted file system
+ * TODO: Add support for unmounting more than
+ * one filesystem.
+ * @return success/error code
+ */
+int unifycr_unmount(void)
+{
+    /************************
+     * tear down connection to server
+     ************************/
+
+    /* detach from shared memory regions */
+    unifycr_shm_free(shm_req_name,  shm_req_size,  &shm_req_buf);
+    unifycr_shm_free(shm_recv_name, shm_recv_size, &shm_recv_buf);
+
+    /* close socket to server */
+    if (client_sockfd >= 0) {
+        errno = 0;
+        int close_rc = close(client_sockfd);
+        if (close_rc != 0) {
+            LOGERR("Failed to close socket to server rc=%d (%s)",
+                errno, strerror(errno));
+        }
+    }
+
+    /* invoke unmount rpc to tell server we're disconnecting */
+    printf("calling unmount\n");
+    int ret = unifycr_client_unmount_rpc_invoke(&unifycr_rpc_context);
+
+    /* free resources allocated in client_rpc_init */
```

Comment:
@adammoody so why is there a rpc finalize and a unifycr_finalize that is not an rpc? I see that is true for init too. Is there a reason they shouldn't be one function?

Reply:
We can look at merging the rpc_init into the unifycr_init too, but I'll do that in a new PR since this is ready to merge.